### PR TITLE
Adapted to akka-persistence-query with tags column

### PR DESCRIPTION
**Pull request by Horusiath**
This is a new implementation of Akka.Persistence.SqlServer. It's compatible with incoming Akka.Persitence.Query plugin.

TODO:

- change dependencies from nighty to standard nuget once 1.1 will be released

### CODE REVIEW

File: src/Akka.Persistence.SqlServer.Tests/Query/SqlServerEventsByPersistenceIdSpec.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="SqliteEventsByPersistenceIdSpec.cs" company="Akka.NET Project">
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/Query/SqlServerEventsByPersistenceIdSpec.cs]: Sqlite

File: src/Akka.Persistence.SqlServer/Journal/SqlServerQueryExecutor.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Data.Common;
+using System.Data.SqlClient;
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer/Journal/SqlServerQueryExecutor.cs]: No copyrights

File: README.md [hunk 3]
```diff
@@ -119,7 +120,13 @@ class MyCustomSqlServerJournal: Akka.Persistence.SqlServer.Journal.SqlServerJour
 
 ### Migration
 
-#### From 1.0.6
+#### From 1.0.8 to 1.1.0
+```SQL
+ALTER TABLE {your_journal_table_name} ADD COLUMN Tags NVARCHAR(2000) NULL;
```
Comment (alexvaluyskiy) [inline README.md]: How to migrate the data?

File: src/Akka.Persistence.SqlServer.Tests/Query/SqlServerAllPersistenceIdsSpec.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            akka.persistence.journal.sql-server {{
+                class = ""Akka.Persistence.SqlServer.Journal.SqlServerJournal, Akka.Persistence.SqlServer""
+                plugin-dispatcher = ""akka.actor.default-dispatcher""
+                table-name = EventJournal
+                schema-name = dbo
+                auto-initialize = on
+                connection-string-name = ""TestDb""
+                refresh-interval = 1s
+            }}
+            akka.test.single-expect-default = 10s")
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/Query/SqlServerAllPersistenceIdsSpec.cs]: We should use bigger timeouts only for debugging

File: src/Akka.Persistence.SqlServer/Snapshot/SqlServerSnapshotStore.cs [hunk 5]
```diff
@@ -1,30 +1,30 @@
...
+        protected readonly SqlServerPersistence Extension = SqlServerPersistence.Get(Context.System);
+        public SqlServerSnapshotStore(Config config) : base(config)
         {
-            _extension = SqlServerPersistence.Get(Context.System);
-            QueryBuilder = new SqlServerSnapshotQueryBuilder(_extension.SnapshotSettings);
-            QueryMapper = new SqlServerQueryMapper(Context.System.Serialization);
+            var sqlConfig = config.WithFallback(Extension.DefaultSnapshotConfig);
+            QueryExecutor = new SqlServerQueryExecutor(new QueryConfiguration(
+                schemaName: config.GetString("schema-name"),
+                snapshotTableName: "Snapshot",
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer/Snapshot/SqlServerSnapshotStore.cs]: we should read it from config

File: src/Akka.Persistence.SqlServer.Tests/packages.config [hunk 6]
```diff
@@ -1,14 +1,28 @@
...
+  <package id="Reactive.Streams" version="1.0.1" targetFramework="net45" />
+  <package id="System.Collections" version="4.0.0" targetFramework="net45" />
   <package id="System.Collections.Immutable" version="1.1.36" targetFramework="net45" />
+  <package id="System.Diagnostics.Debug" version="4.0.0" targetFramework="net45" />
+  <package id="System.Globalization" version="4.0.0" targetFramework="net45" />
+  <package id="System.Linq" version="4.0.0" targetFramework="net45" />
+  <package id="System.Resources.ResourceManager" version="4.0.0" targetFramework="net45" />
+  <package id="System.Runtime" version="4.0.0" targetFramework="net45" />
+  <package id="System.Runtime.Extensions" version="4.0.0" targetFramework="net45" />
+  <package id="System.Threading" version="4.0.0" targetFramework="net45" />
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/packages.config]: You should remove this redundant packages from NetCore

Comment (Horusiath) [inline src/Akka.Persistence.SqlServer.Tests/packages.config]: Which of them are redundant? They are all added by System.Collections.Immutable dependency.

File: README.md (hunk 3 above)
Comment (Horusiath) [inline README.md]: Could you clarify? Tags are new (nullable) column, so no migration is needed.

File: src/Akka.Persistence.SqlServer.Tests/packages.config (hunk 6 above)
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/packages.config]: They was added in System.Collections.Immutable 1.1.37, not in 1.1.36. That's why this lib was downgraded in Akka repo

File: README.md (hunk 3 above)
Comment (alexvaluyskiy) [inline README.md]: Timestamp was DATETIME2 column and became BigInt

**Comment by Horusiath** [comment]
Made some fixes and updates accordingly to @alexvaluyskiy notes. Also updated dependencies from akka nightly builds to regular nuget. If tests will pass, this should be ready to go.

File: src/Akka.Persistence.SqlServer.Tests/Query/SqlServerEventsByPersistenceIdSpec.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+using Akka.Configuration;
+using Akka.Persistence.Query.Sql;
+using Akka.Persistence.Sql.TestKit;
+using Akka.Util.Internal;
+using Xunit.Abstractions;
+
+namespace Akka.Persistence.SqlServer.Tests.Query
+{
+    public class SqliteEventsByPersistenceIdSpec : EventsByPersistenceIdSpec
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/Query/SqlServerEventsByPersistenceIdSpec.cs]: Still Sqlite

File: src/Akka.Persistence.SqlServer.Tests/Akka.Persistence.SqlServer.Tests.csproj [hunk 8]
```diff
@@ -78,6 +112,10 @@
       <HintPath>..\packages\Newtonsoft.Json.7.0.1\lib\net45\Newtonsoft.Json.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="Reactive.Streams, Version=1.0.1.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/Akka.Persistence.SqlServer.Tests.csproj]: Maybe we should reference Reactive.Streams 1.0.0-RC1?

**Comment by alexvaluyskiy** [comment]
@Horusiath  why you put two JournalEntry classes in SqlCommon?
https://github.com/akkadotnet/akka.net/blob/dev/src/contrib/persistence/Akka.Persistence.Sql.Common/Journal/JournalDbEngine.cs#L26
https://github.com/akkadotnet/akka.net/blob/dev/src/contrib/persistence/Akka.Persistence.Sql.Common/Journal/Messages.cs#L18

**Comment by alexvaluyskiy** [comment]
@Horusiath Oh I understood. We have redundant files in the repo

**Comment by alexvaluyskiy** [comment]
@Horusiath this test should be unskipped
https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Persistence.TestKit/Snapshot/SnapshotStoreSpec.cs#L214

**Comment by Horusiath** [comment]
It looks like the reason behind the failure is:

```
Method not found: 'System.__Canon Akka.TestKit.TestKitBase.Within(System.TimeSpan, System.TimeSpan, System.Func`1<System.__Canon>, System.String)'.

   at Akka.Streams.TestKit.TestSubscriber.ManualProbe`1.Within[TOther](TimeSpan max, Func`1 func)
   at Akka.Persistence.Sql.TestKit.AllPersistenceIdsSpec.Sql_query_AllPersistenceIds_should_find_existing_persistence_ids()
```

Which points to the core repo (and akka streams testkit). /cc @Silv3rcircl3

**Comment by alexvaluyskiy** [comment]
@Horusiath @Silv3rcircl3  it's a new bug, which was introduced in 1.1.1 here
https://github.com/akkadotnet/akka.net/pull/2177

It works fine in 1.1.0

File: README.md [hunk 9]
```diff
@@ -76,16 +76,17 @@ akka.persistence{
...
 CREATE TABLE {your_journal_table_name} (
   PersistenceID NVARCHAR(255) NOT NULL,
   SequenceNr BIGINT NOT NULL,
-  Timestamp DATETIME2 NOT NULL,
+  Timestamp BIGINT NOT NULL,
   IsDeleted BIT NOT NULL,
   Manifest NVARCHAR(500) NOT NULL,
-  Payload VARBINARY(MAX) NOT NULL
+  Payload VARBINARY(MAX) NOT NULL,
+	Tags NVARCHAR(2000) NULL
```
Comment (alexvaluyskiy) [inline README.md]: change to NVARCHAR(100)

File: src/Akka.Persistence.SqlServer.Tests/Query/SqlServerAllPersistenceIdsSpec.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+using Akka.Configuration;
+using Akka.Persistence.Query.Sql;
+using Akka.Persistence.Sql.TestKit;
+using Akka.Util.Internal;
+using Xunit.Abstractions;
+
+namespace Akka.Persistence.SqlServer.Tests.Query
+{
+    public class SqlServerAllPersistenceIdsSpec : AllPersistenceIdsSpec
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/Query/SqlServerAllPersistenceIdsSpec.cs]: We should add [Collection("SqlServerSpec")] attribute to disable parallelization

File: src/Akka.Persistence.SqlServer.Tests/Query/SqlServerAllPersistenceIdsSpec.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+                plugin-dispatcher = ""akka.actor.default-dispatcher""
+                table-name = EventJournal
+                schema-name = dbo
+                auto-initialize = on
+                connection-string-name = ""TestDb""
+                refresh-interval = 1s
+            }}")
+            .WithFallback(SqlReadJournal.DefaultConfiguration());
+
+        public SqlServerAllPersistenceIdsSpec(ITestOutputHelper output) : base(Config(Counter.GetAndIncrement()), output)
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/Query/SqlServerAllPersistenceIdsSpec.cs]: We don't need AtomicCounter here

File: src/Akka.Persistence.SqlServer.Tests/Query/SqlServerAllPersistenceIdsSpec.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+            akka.persistence.journal.plugin = ""akka.persistence.journal.sql-server""
+            akka.persistence.journal.sql-server {{
+                class = ""Akka.Persistence.SqlServer.Journal.SqlServerJournal, Akka.Persistence.SqlServer""
+                plugin-dispatcher = ""akka.actor.default-dispatcher""
+                table-name = EventJournal
+                schema-name = dbo
+                auto-initialize = on
+                connection-string-name = ""TestDb""
+                refresh-interval = 1s
+            }}")
```
Comment (alexvaluyskiy) [inline src/Akka.Persistence.SqlServer.Tests/Query/SqlServerAllPersistenceIdsSpec.cs]: Works better with akka.test.single-expect-default = 5s

**Comment by alexvaluyskiy** [comment]
@Horusiath You should update to Akka 1.1.1 and your exception will dissapear

**Comment by Horusiath** [comment]
Everything is passing locally, both with resharper and by using FAKE build scripts. I don't know WTF at this moment.

**Comment by alexvaluyskiy** [comment]
Just increase the timeouts (to 10 sec) in Query tests and create an issue for that. Some parts of Streams slower that it should be at the moment

**Comment by alexvaluyskiy** [comment]
Akka Jvm also has 10s
https://github.com/akka/akka/blob/master/akka-persistence-query/src/test/scala/akka/persistence/query/journal/leveldb/EventsByPersistenceIdSpec.scala

**Comment by Aaronontheweb** [comment]
> Everything is passing locally, both with resharper and by using FAKE build scripts. I don't know WTF at this moment.

Possible that there can be a concurrency bug here, given that the degree of parallelism on the build server is going to be lower than your local machine.

**Comment by Aaronontheweb** [comment]
Reverted this. Will need to re-submit under a new PR

**Pull request merged**